File: solutions/python/normalized_pathnames.py
import sys


# @include
def shortest_equivalent_path(path):
    if not path:
        raise ValueError('Empty string is not a valid path.')

    path_names = []  # Uses list as a stack.
    # Special case: starts with '/', which is an absolute path.
    if path[0] == '/':
        path_names.append('/')

    for token in (token for token in path.split('/') if token not in ['.', '']):
        if token == '..':
            if not path_names or path_names[-1] == '..':
                path_names.append(token)
            else:
                if path_names[-1] == '/':
                    raise ValueError('Path error')
                path_names.pop()
        else:  # Must be a name.
            path_names.append(token)

    result = '/'.join(path_names)
    return result[result.startswith('//'):]  # Avoid starting '//'.
# @exclude


def main():
    assert shortest_equivalent_path('123/456') == '123/456'
    assert shortest_equivalent_path('/123/456') == '/123/456'
    assert shortest_equivalent_path('usr/lib/../bin/gcc') == 'usr/bin/gcc'
    assert shortest_equivalent_path('./../') == '..'
    assert shortest_equivalent_path('../../local') == '../../local'
    assert shortest_equivalent_path('./.././../local') == '../../local'
    assert shortest_equivalent_path('/foo/../foo/./../') == '/'
    try:
        shortest_equivalent_path('/foo.txt')
    except ValueError as e:
        print(e)
    try:
        shortest_equivalent_path('/..')
    except ValueError as e:
        print(e)
    try:
        shortest_equivalent_path('/cpp_name/bin/')
    except ValueError as e:
        print(e)
    assert (shortest_equivalent_path('scripts//./../scripts/awkscripts/././') ==
            'scripts/awkscripts')
    if len(sys.argv) == 2:
        print(shortest_equivalent_path(sys.argv[1]))


if __name__ == '__main__':
    main()

File: solutions/cpp/normalized_pathnames.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::exception;
using std::invalid_argument;
using std::stringstream;
using std::string;
using std::vector;

// @include
string ShortestEquivalentPath(const string& path) {
  if (path.empty()) {
    throw invalid_argument("Empty string is not a valid path.");
  }

  vector<string> path_names;  // Uses vector as a stack.
  // Special case: starts with "/", which is an absolute path.
  if (path.front() == '/') {
    path_names.emplace_back("/");
  }

  stringstream ss(path);
  string token;
  while (getline(ss, token, '/')) {
    if (token == "..") {
      if (path_names.empty() || path_names.back() == "..") {
        path_names.emplace_back(token);
      } else {
        if (path_names.back() == "/") {
          throw invalid_argument("Path error");
        }
        path_names.pop_back();
      }
    } else if (token != "." && token != "") {  // Must be a name.
      path_names.emplace_back(token);
    }
  }

  string result;
  if (!path_names.empty()) {
    result = path_names.front();
    for (int i = 1; i < path_names.size(); ++i) {
      if (i == 1 && result == "/") {  // Avoid starting "//".
        result += path_names[i];
      } else {
        result += "/" + path_names[i];
      }
    }
  }
  return result;
}
// @exclude

int main(int argc, char* argv[]) {
  assert(ShortestEquivalentPath("123/456") == string("123/456"));
  assert(ShortestEquivalentPath("/123/456") == string("/123/456"));
  assert(ShortestEquivalentPath("usr/lib/../bin/gcc") ==
         string("usr/bin/gcc"));
  assert(ShortestEquivalentPath("./../") == string(".."));
  assert(ShortestEquivalentPath("../../local") == string("../../local"));
  assert(ShortestEquivalentPath("./.././../local") == string("../../local"));
  assert(ShortestEquivalentPath("/foo/../foo/./../") == string("/"));
  try {
    ShortestEquivalentPath("/foo.txt");
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  try {
    ShortestEquivalentPath("");
    assert(false);
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  try {
    ShortestEquivalentPath("/..");
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  try {
    ShortestEquivalentPath("/cpp_name/bin/");
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  assert(ShortestEquivalentPath("scripts//./../scripts/awkscripts/././") ==
         string("scripts/awkscripts"));
  if (argc == 2) {
    cout << ShortestEquivalentPath(argv[1]) << endl;
  }
  return 0;
}
